### DISCUSSION TITLE
Capacitor Push Notifications FCM - Android - Notifications don't appear in action bar when app in foreground

### DISCUSSION THREAD

**Discussion by bspcbord** [ionic-team/capacitor#3286] [Q&A] [answered]
<h2>Capacitor Push Notifications FCM - Android - Notifications don't appear in action bar when app in foreground</h2>

<h3>Problem:</h3>
This works as expected in iOS. I am able to register and add notification listeners just fine.  I'm able to log the output of the notification when the app is in the foreground, but I see no notification in the action bar.  If I put the app into the background, the notification appears in the action bar as expected.

I would like to see the notification in the action bar when the app is in the foreground and background.

<h3>Notification Payload sent to FCM</h3>

```
{
   "registration_ids":[
      "myAppId"
   ],
   "priority":10,
   "notification":{
      "title":"GET Message",
      "body":"7.17-8:39",
      "sound":"default"
   }
}
```

<h3>Notification Log from `pushNotificationReceived` listener in app</h3>

```
{
    body: "Test Message"
    data: {}
    id: "someIdValue"
    title: "Test Title"
}
```

<h3>Ionic info</h3>
Ionic:
   Ionic CLI                     : 5.4.16
   Ionic Framework               : @ionic/angular 4.11.10
   @angular-devkit/build-angular : 0.13.9
   @angular-devkit/schematics    : 7.2.4
   @angular/cli                  : 7.3.9
   @ionic/angular-toolkit        : 1.5.1

Capacitor:
   Capacitor CLI   : 2.2.1
   @capacitor/core : 2.2.1

Cordova:
   Cordova CLI       : 9.0.3 (cordova-lib@9.0.2)
   Cordova Platforms : none
   Cordova Plugins   : no whitelisted plugins (0 plugins total)

Utility:
   cordova-res : not installed
   native-run  : not installed

System:
   NodeJS : v10.15.3 (C:\Program Files\nodejs\node.exe)
   npm    : 6.3.0
   OS     : Windows 10

<h3>Any guidance would be greatly appreciated</h3>

**Comment by bspcbord** [comment, marked as answer]
I think I may have found the reason for this in the docs:
https://firebase.google.com/docs/cloud-messaging/android/receive#handling_messages

App in foreground will only receive messages in the event listener, not the system tray.

**Comment by adnan-hd** [comment]
@bspcbord , Hello, were you able to resolve the issue using Capacitor-pushNotificationReceived() ? 
I do not want to use FCM for notifications, as the capacitor plugin works well in other platforms for this scenario.

**Comment by alex-migwi** [comment]
When in the foreground on receiving the notification get its payload and create a local notification and schedule it to be shown immediately..... 
This is a sample from my code which should give you enough ideas on what to do.

```typescript
     // Only called when app in foreground
    Plugins.PushNotifications.addListener('pushNotificationReceived', (notification) => {
      console.log('Notification received:', notification);
      this.doLocalNotif(notification);
 
      Plugins.LocalNotifications.addListener(
        'localNotificationActionPerformed',
        (event: LocalNotificationActionPerformed) => {
          this.router.navigateByUrl('messages');
        }
      );
    });

    // Handle foreground notification as a local one
   private doLocalNotif(notification) {
    Plugins.LocalNotifications.schedule({
      notifications: [
        {
          title: notification.title,
          body: notification.body,
          id: Date.now(),
          attachments: [
            {
              id: 'site',
              url: notification.data.imageUrl,
            },
          ],
          actionTypeId: '',
        },
      ],
    });
  }
```

**Reply by SalahAdDin** [reply]
Is there any way to use that router with **IonicReact**?

**Reply by alex-migwi** [reply]
I work primarily with angular but i'm sure you can...
This article talks about programmatic routing in react may it can help you along [https://ui.dev/react-router-v4-programmatically-navigate/](url)

**Reply by vladern** [reply]
Thank you very much, indeed this seems to be the way to handle notifications in the foreground

**Comment by kiranrayala77** [comment]
Plugins.PushNotifications.addListener('pushNotificationReceived', (notification) => {
      console.log('Notification received:', notification);
      this.doLocalNotif(notification);
 
      Plugins.LocalNotifications.addListener(
        'localNotificationActionPerformed',
        (event: LocalNotificationActionPerformed) => {
          this.router.navigateByUrl('messages');
        }
      );
    });

This is not triggered in my react using capacitor code when the app is in foreground
